lightning/src/util/persist.rs

```diff
@@ -249,10 +259,966 @@ where
+						monitor_name.as_str(),
+						update_name.as_str(),
+						e
+					);
+					io::Error::new(io::ErrorKind::Other, "Monitor update failed")
+				})?;
+			}
+			// ...push the result into the return vec
+			res.push((block_hash, monitor))
+		}
+		Ok(res)
+	}
+
+	/// Read a channel monitor.
+	fn read_monitor(
+		&self, monitor_name: &MonitorName,
+	) -> io::Result<(BlockHash, ChannelMonitor<<SP::Target as SignerProvider>::Signer>)> {
+		let outpoint: OutPoint = monitor_name.try_into()?;
+		let mut monitor_cursor = io::Cursor::new(self.kv_store.read(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+			monitor_name.as_str(),
+		)?);
+		// Discard the sentinel bytes if found.
+		if monitor_cursor.get_ref().starts_with(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL) {
+			monitor_cursor.set_position(MONITOR_UPDATING_PERSISTER_PREPEND_SENTINEL.len() as u64);
+		}
+		match <(BlockHash, ChannelMonitor<<SP::Target as SignerProvider>::Signer>)>::read(
+			&mut monitor_cursor,
+			(&*self.entropy_source, &*self.signer_provider),
+		) {
+			Ok((blockhash, channel_monitor)) => {
+				if channel_monitor.get_funding_txo().0.txid != outpoint.txid
+					|| channel_monitor.get_funding_txo().0.index != outpoint.index
+				{
+					log_error!(
+						self.logger,
+						"ChannelMonitor {} was stored under the wrong key!",
+						monitor_name.as_str()
+					);
+					return Err(io::Error::new(
+						io::ErrorKind::InvalidData,
+						"ChannelMonitor was stored under the wrong key",
+					));
+				}
+				Ok((blockhash, channel_monitor))
+			}
+			Err(e) => {
+				log_error!(
+					self.logger,
+					"Failed to read ChannelMonitor {}, reason: {}",
+					monitor_name.as_str(),
+					e,
+				);
+				Err(io::Error::new(io::ErrorKind::InvalidData, "Failed to read ChannelMonitor"))
+			}
+		}
+	}
+
+	/// Read a channel monitor update.
+	fn read_monitor_update(
+		&self, monitor_name: &MonitorName, update_name: &UpdateName,
+	) -> io::Result<ChannelMonitorUpdate> {
+		let update_bytes = self.kv_store.read(
+			CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE,
+			monitor_name.as_str(),
+			update_name.as_str(),
+		)?;
+		ChannelMonitorUpdate::read(&mut io::Cursor::new(update_bytes)).map_err(|e| {
+			log_error!(
+				self.logger,
+				"Failed to read ChannelMonitorUpdate {}/{}/{}, reason: {}",
+				CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE,
+				monitor_name.as_str(),
+				update_name.as_str(),
+				e,
+			);
+			io::Error::new(io::ErrorKind::InvalidData, "Failed to read ChannelMonitorUpdate")
+		})
+	}
+
+	/// Cleans up stale updates for all monitors.
+	///
+	/// This function works by first listing all monitors, and then for each of them, listing all
+	/// updates. The updates which have an update_id less than or equal to than the stored monitor
+	/// are deleted. The deletion can either be lazy or non-lazy based on the `lazy` flag; this will
+	/// be passed to [`KVStore::remove`].
```

Comment:
Want to note that I thought about adding a note here about memory usage, but I think it's okay, there's no footgun here. I think it's something like 1MB per 31,250 updates (MonitorName is 32B and holds both the u64 and String representation). Pagination in listing is also a problem when this gets big, but should just be in the `KVStore` backend and never matter here.